File: src/llm/provider.ts
import {
  createOpenRouter,
  type OpenRouterChatSettings,
} from "@openrouter/ai-sdk-provider";

export const WORKHORSE_ID = process.env.WORKHORSE_MODEL ?? "google/gemini-2.5-flash";
export const PREMIUM_ID = process.env.PREMIUM_MODEL ?? "anthropic/claude-sonnet-4.6";

let provider: ReturnType<typeof createOpenRouter> | undefined;

// Lazy so importing this module (e.g. to print CLI usage) doesn't require the key.
function openrouter() {
  if (!provider) {
    const apiKey = process.env.OPENROUTER_API_KEY;
    if (!apiKey) {
      throw new Error(
        "OPENROUTER_API_KEY is not set. Get a key at https://openrouter.ai/keys and provide it via the environment (e.g. a .env file, or `OPENROUTER_API_KEY=... nr research <SYMBOL>`).",
      );
    }
    provider = createOpenRouter({ apiKey });
  }
  return provider;
}

// `usage.include` turns on OpenRouter usage accounting so each response carries
//   providerMetadata.openrouter.usage = { cost, promptTokensDetails.cachedTokens, ... }
// which is how W1 reports real $ and cache-hit tokens.
// `data_collection: "deny"` keeps financial prompts off providers that log/train.
const BASE: OpenRouterChatSettings = {
  usage: { include: true },
  provider: { data_collection: "deny" },
};

function merge(extra?: OpenRouterChatSettings): OpenRouterChatSettings {
  return {
    ...BASE,
    ...extra,
    provider: { ...BASE.provider, ...extra?.provider },
  };
}

// Cheap, high-volume calls (research draft, discovery). Fallbacks allowed for
// resilience; callers may pass { provider: { sort: "price" } } to chase the floor.
export const workhorse = (extra?: OpenRouterChatSettings) =>
  openrouter()(WORKHORSE_ID, merge(extra));

// Quality-critical calls (the critic chain). `allow_fallbacks: false` makes
// routing sticky so the cached prefix lands on the same provider endpoint across
// calls 2–4 (W2) — Anthropic prompt caching only pays off on cache hits.
export const premium = (extra?: OpenRouterChatSettings) =>
  openrouter()(
    PREMIUM_ID,
    merge({ ...extra, provider: { allow_fallbacks: false, ...extra?.provider } }),
  );

File: app/llm/provider.py
"""Pydantic AI / OpenRouter model factories.

Lazy: importing this module (e.g. to print CLI usage) must not require the key.

- workhorse: cheap, high-volume calls (research draft, discovery). May chase the
  price floor via `sort: "price"`.
- premium: quality-critical calls (the critic chain). `allow_fallbacks: False`
  makes routing sticky so the cached prefix lands on the same provider endpoint
  across calls — Anthropic prompt caching only pays off on cache hits.

`data_collection: "deny"` keeps financial prompts off providers that log/train.
`usage.include` turns on OpenRouter usage accounting for real $ + cached tokens.
"""

from __future__ import annotations

from pydantic_ai.models.openrouter import OpenRouterModel, OpenRouterModelSettings
from pydantic_ai.providers.openrouter import OpenRouterProvider

from ..config import OPENROUTER_API_KEY, PREMIUM_MODEL, WORKHORSE_MODEL

_provider: OpenRouterProvider | None = None


def _get_provider() -> OpenRouterProvider:
    global _provider
    if _provider is None:
        if not OPENROUTER_API_KEY:
            raise RuntimeError(
                "OPENROUTER_API_KEY is not set. Get a key at "
                "https://openrouter.ai/keys and provide it via the environment "
                "(e.g. a .env file)."
            )
        _provider = OpenRouterProvider(api_key=OPENROUTER_API_KEY)
    return _provider


def workhorse_model() -> OpenRouterModel:
    return OpenRouterModel(WORKHORSE_MODEL, provider=_get_provider())


def premium_model() -> OpenRouterModel:
    return OpenRouterModel(PREMIUM_MODEL, provider=_get_provider())


def workhorse_settings(*, price_sort: bool = False) -> OpenRouterModelSettings:
    provider: dict = {"data_collection": "deny"}
    if price_sort:
        provider["sort"] = "price"
    return OpenRouterModelSettings(openrouter_provider=provider, openrouter_usage={"include": True})


def premium_settings() -> OpenRouterModelSettings:
    return OpenRouterModelSettings(
        openrouter_provider={"data_collection": "deny", "allow_fallbacks": False},
        openrouter_usage={"include": True},
    )
